### ISSUE TITLE
Incorrect sugestion for foldr

### ISSUE THREAD

**Issue by basile-henry** [ndmitchell/hlint#793]
The following code:

```haskell
combinations :: [a] -> [(a, a)]
combinations [] = []
combinations (x:xs) = map ((,) x) xs ++ combinations xs
```

Gives me the following `hlint` suggestions:

```sh
Test.hs:15:1: Suggestion: Use foldr
Found:
  combinations [] = []
  combinations (x : xs) = map ((,) x) xs ++ combinations xs
Perhaps:
  combinations xs = foldr (\ x -> (++) (map ((,) x) xs)) [] xs

Test.hs:16:27: Warning: Use tuple-section
Found:
  ((,) x)
Perhaps:
  (x,)
Note: may require `{-# LANGUAGE TupleSections #-}` adding to the top of the file
```

The second suggestion about tuple section is fine, the first is not!

Even though I can see the suggested code is not equivalent to what I wrote, let's have QuickCheck confirm that for me:

```haskell
import Test.QuickCheck

combinations :: [a] -> [(a, a)]
combinations [] = []
combinations (x:xs) = map ((,) x) xs ++ combinations xs

combinations2 :: [a] -> [(a, a)]
combinations2 xs = foldr (\ x -> (++) (map ((,) x) xs)) [] xs

main = quickCheck $ \xs -> combinations xs == combinations2 (xs :: [Int])
```

```sh
$ runhaskell Test.hs
*** Failed! Falsified (after 2 tests and 1 shrink):
[0]
```

This is using `hlint` version `2.2.4` (installed from nixpkgs).

<hr/>

I think we need to have an extra check in this suggestion that `xs` is not used in the accumulation function given to `foldr`, if it is then the suggestion is wrong.

**Comment by ndmitchell** [comment]
Thanks for the report! You are indeed right, an extra check that `xs` isn't used is required. Previously I had a check that neither `x` nor `xs` were used in `++`. That's neither necessary (it's fine if `x` occurs) nor sufficient (`xs` can't occur in the bit to the left of the `++` either). I've fixed that, and will include the fix in the next release, probably in a few weeks time.
